File: app/src/main/java/com/expensivebelly/dagger2retaingraph/fragment/master/di/MasterModule.java
package com.expensivebelly.dagger2retaingraph.fragment.master.di;

import com.expensivebelly.dagger2retaingraph.activity.di.ActivityScope;
import com.expensivebelly.dagger2retaingraph.fragment.master.MasterPresenter;

import java.util.concurrent.TimeUnit;

import dagger.Module;
import dagger.Provides;
import io.reactivex.Observable;
import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

@Module
class MasterModule {

    @Provides
    @ActivityScope
    static MasterPresenter provideMasterPresenter(Single<Long> timeWaster) {
        return new MasterPresenter(timeWaster);
    }

    @Provides
    static Single<Long> provideTimeWaster() {
        long period = 5;
        return Observable.interval(period, TimeUnit.SECONDS)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .cache()
                .first(period);
    }
}

File: app/src/main/java/com/expensivebelly/dagger2retaingraph/fragment/master/di/MasterModule.kt
package com.expensivebelly.dagger2retaingraph.fragment.master.di

import com.expensivebelly.dagger2retaingraph.activity.di.ActivityScope
import com.expensivebelly.dagger2retaingraph.fragment.master.MasterPresenter
import dagger.Module
import dagger.Provides
import io.reactivex.rxjava3.core.Single
import java.util.concurrent.TimeUnit

@Module
class MasterModule {

    @Provides
    @ActivityScope
    fun provideMasterPresenter(timeWaster: Single<Long>): MasterPresenter {
        return MasterPresenter(timeWaster)
    }

    @Provides
    fun provideTimeWaster(): Single<Long> {
        val period: Long = 5
        return Single.just(period)
                .toObservable().replay(1).refCount(1, TimeUnit.SECONDS)
                .delay(period, TimeUnit.SECONDS)
                .firstOrError()
    }
}